src/Dom/Document.php

```diff
@@ -1099,6 +1183,54 @@ private function convertAmpBindAttributes($html)
         return (null !== $converted) ? $converted : $html;
     }
 
+    /**
+     * Convert AMP bind-attributes back to their original syntax.
+     *
+     * This is not guaranteed to produce the exact same result as the initial markup, as it is more of a best guess.
+     * It can end up replacing the wrong attributes if the initial markup had inconsistent styling, mixing both syntaxes
+     * for the same attribute. In either case, it will always produce working markup, so this is not that big of a deal.
+     *
+     * @see convertAmpBindAttributes() Reciprocal function.
+     * @link https://www.ampproject.org/docs/reference/components/amp-bind
+     *
+     * @param string $html HTML with amp-bind attributes converted.
+     * @return string HTML with amp-bind attributes restored.
+     */
+    public function restoreAmpBindAttributes($html)
+    {
+        if ($this->options[self::OPTION_AMP_BIND_SYNTAX] === self::AMP_BIND_SYNTAX_DATA_ATTRIBUTE) {
+            // All amp-bind attributes should remain in their converted data attribute form.
+            return $html;
+        }
+
+        if (
+            $this->options[self::OPTION_AMP_BIND_SYNTAX] === self::AMP_BIND_SYNTAX_AUTO
+            &&
+            empty($this->convertedAmpBindAttributes)
+        ) {
+            // Only previously converted amp-bind attributes should be restored, but none were converted.
+            return $html;
+        }
+
+        // Depending on options, restore previously converted or all amp-bind attributes to square bracket syntax.
+        $pattern = $this->options[self::OPTION_AMP_BIND_SYNTAX] === self::AMP_BIND_SYNTAX_AUTO
+            ? sprintf(
+                '#%s(%s)#i',
+                self::AMP_BIND_DATA_ATTR_PREFIX,
+                implode('|', array_unique($this->convertedAmpBindAttributes))
+            )
+            : sprintf(
+                '#%s([a-z0-9-_]+)#i', // TODO: Check what exact regex makes sense here.
+                self::AMP_BIND_DATA_ATTR_PREFIX
+            );
```

Comment:
Don't the patterns here run the risk of unintentional conversion? For example, if I write a post about amp-bind and I say:

> You have two options: use the `[text]` attribute or the `data-amp-bind-text` attribute.

At present this will result in the unexpected:

> You have two options: use the `[text]` attribute or the `[text]` attribute.

In order to safely convert back to the bracketed syntax, we'd need to use the same logic in `convertAmpBindAttributes` which uses the `AMP_BIND_START_TAGS_PATTERN` regex to first match all tags which contain an attribute in question, and then in the replacement callback do the actual replacements.

Reply:
Yes, indeed, this needs to become way more robust.

Reply:
I think this should now work as expected (see added tests with latest commits).

The regex I ended up with looks overly complex, so happy to hear feedback if I'm missing something obvious.